pygmt/datatypes.py

```diff
@@ -0,0 +1,221 @@
+"""
+Wrappers for GMT data types.
+"""
+import ctypes as ctp
+
+import numpy as np
+
+
+class GMT_DATASET(ctp.Structure):  # pylint: disable=invalid-name,too-few-public-methods
```

Comment:
The `GMT_DATASET` object is difficult to use (see the docstrings and the `to_vectors` method) and should not be exposed to our end users. So maybe we should rename it to something like `_GMT_DATASET`?

On the other hand, we may want to provide a Pythonic data structure to represent GMT's `GMT_DATASET` object, like what GMT.jl does (https://www.generic-mapping-tools.org/GMT.jl/dev/types/#Dataset-type). The new Pythonic data structure is necessary to let PyGMT support multi-segment datasets but it can be done in future PRs after more discussions.

Reply:
> The `GMT_DATASET` object is difficult to use (see the docstrings and the `to_vectors` method) and should not be exposed to our end users. So maybe we should rename it to something like `_GMT_DATASET`?

Instead of naming the class `_GMT_DATASET`, you could also name the file `_datatypes.py` to mark it private.

> On the other hand, we may want to provide a Pythonic data structure to represent GMT's `GMT_DATASET` object, like what GMT.jl does (https://www.generic-mapping-tools.org/GMT.jl/dev/types/#Dataset-type). The new Pythonic data structure is necessary to let PyGMT support multi-segment datasets but it can be done in future PRs after more discussions.

Let's discuss this in a separate issue. I'd much prefer that we don't reinvent another `pandas.DataFrame` or `geopandas.GeoDataFrame` data structure for Python users (but having an internal `GMT_DATASET` class is ok). Also we should be looking at upcoming standards like the [Python dataframe interchange protocol](https://data-apis.org/dataframe-protocol/latest/index.html) and [GeoArrow](https://github.com/geoarrow/geoarrow) if we want a data structure that is reusable across the Python ecosystem.

Reply:
> > The `GMT_DATASET` object is difficult to use (see the docstrings and the `to_vectors` method) and should not be exposed to our end users. So maybe we should rename it to something like `_GMT_DATASET`?
> 
> Instead of naming the class `_GMT_DATASET`, you could also name the file `_datatypes.py` to mark it private.

`_datatypes.py` sounds good. Will do it later.

> > On the other hand, we may want to provide a Pythonic data structure to represent GMT's `GMT_DATASET` object, like what GMT.jl does (https://www.generic-mapping-tools.org/GMT.jl/dev/types/#Dataset-type). The new Pythonic data structure is necessary to let PyGMT support multi-segment datasets but it can be done in future PRs after more discussions.
> 
> Let's discuss this in a separate issue. I'd much prefer that we don't reinvent another `pandas.DataFrame` or `geopandas.GeoDataFrame` data structure for Python users (but having an internal `GMT_DATASET` class is ok). Also we should be looking at upcoming standards like the [Python dataframe interchange protocol](https://data-apis.org/dataframe-protocol/latest/index.html) and [GeoArrow](https://github.com/geoarrow/geoarrow) if we want a data structure that is reusable across the Python ecosystem.

Yes, definitely needs more discussions to support multi-segment data files.